### PR TITLE
12 validate taxonomy

### LINKED ISSUE #12: Validate DBpedia taxonomisation
Assessing between methods

### PR DESCRIPTION
**Pull request by ampudia19**
---
# Taxonomy validation (Phase 1)
# Description
The first set of validation activities includes visualisation tools as well as two approaches to labelling topics (entity labelling and weighted journal labelling). Plots include sunbursts and a first set of histograms. For details, see the README file in `pipeline/taxonomy_validation`.

Fixes #12 

# Checklist:

- [X] I have refactored my code out from `notebooks/`
- [X] I have checked the code runs
- [X] I have tested the code
- [X] I have run `pre-commit` and addressed any issues not automatically fixed
- [X] I have merged any new changes from `dev`
- [X] I have documented the code
  - [X] Major functions have docstrings
  - [X] Appropriate information has been added to `README`s
- [X] I have explained this PR above
- [] I have requested a code review

### CODE REVIEW

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        )
+    elif args.taxonomy_class == "semantic_kmeans":
+        taxonomy = pipe(
+            "kmeans_strict_imb",
+            get_semantic_taxonomy,
+            lambda df: df.rename(columns={"tag": "Entity"})
+            if "tag" in df.columns
+            else df,
+            lambda df: df.set_index("Entity"),
+            lambda df: df.rename(
```
Comment (ampudia19) [inline dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: These are included here because I zero-indexed my levels (whereas Emily one-indexed them). I'm changing it in branch 9_, so this should eventually be reverted.

File: dap_aria_mapping/getters/taxonomies.py [hunk 2]
```diff
@@ -48,3 +48,15 @@ def get_test_cooccurrence_taxonomy() -> pd.DataFrame:
         "outputs/test_community_detection_clusters.parquet",
         download_as="dataframe",
     )
+
+
+def get_semantic_taxonomy(cluster_object: str = "centroids") -> pd.DataFrame:
```
Comment (emily-bicks) [inline dap_aria_mapping/getters/taxonomies.py:53]: Missing docstring for what "cluster_object" is and the available options

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+from dap_aria_mapping.getters.taxonomies import (
+    get_cooccurrence_taxonomy,
+    get_semantic_taxonomy,
+)
+from dap_aria_mapping.getters.openalex import get_openalex_works, get_openalex_entities
+from dap_aria_mapping.utils.semantics import get_sample, filter_entities
+from dap_aria_mapping.utils.validation import *
+
+
+def dataframe_to_histogram_df(
```
Comment (emily-bicks) [inline dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:22]: I think this should be in a more general util - I think we will use the taxonomy distributions for more than just plotting histograms

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        "--n_articles",
+        type=int,
+        help="The number of articles to use. If -1, all articles are used.",
+    )
+
+    parser.add_argument("--save", help="Whether to save the plot.", action="store_true")
+
+    args = parser.parse_args()
+
+    logger.info("Loading data - taxonomy")
```
Comment (emily-bicks) [inline dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:89]: Should we also have the taxonomies as command line arguments? I can see us wanting to run these for specific taxonomies or for all taxonomies

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    )
+
+    semantic_kmeans_taxonomy_labelled = build_labelled_taxonomy(
+        semantic_kmeans_taxonomy, journal_entities, entity_counts, entity_journal_counts
+    )
+
+    levels = args.level if isinstance(args.level, list) else list(args.level)
+
+    for level in levels:
+        chart = alt.vconcat(
```
Comment (emily-bicks) [inline dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:142]: you can format the altair chart using the nesta theme with 

```
from nesta_ds_utils.viz.altair import formatting, saving
formatting.setup_theme()
```

and save it using 
`saving.save(chart, name='test_figure', path='outputs/figures')
`

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(
+        description="Make histograms of a given taxonomy level.",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "--taxonomy_class",
```
Comment (emily-bicks) [inline dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: should we just shorten this to 'taxonomy'?

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    )
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Body=pickle.dumps(labels),
+        Bucket=BUCKET_NAME,
+        Key=title,
+    )
+
+
+if __name__ == "__main__":
```
Comment (emily-bicks) [inline dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: Would suggest naming this script something a bit more descriptive than `make_labels.py`. Something like `generate_journal_distributions.py` or something like that?

File: dap_aria_mapping/utils/semantics.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
+import numpy as np
```
Comment (emily-bicks) [inline dap_aria_mapping/utils/semantics.py]: These are really useful - would name the script entity_selection.py as I could see these utils being used for a variety of use cases, not just the semantic models

File: dap_aria_mapping/utils/validation.py [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+
+
+def build_labelled_taxonomy(
+    df: pd.DataFrame,
+    journal_entities: Dict[str, Sequence[str]],
+    entity_counts: Dict[str, int],
+    entity_journal_counts: Dict[str, Dict[str, int]],
+    levels: int = 3,
+) -> pd.DataFrame:
+    """Builds a labelled taxonomy from a taxonomy dataframe, a
```
Comment (emily-bicks) [inline dap_aria_mapping/utils/validation.py]: Would be helpful to be a bit more descriptive about what the 'labeled taxonomy' is

**Review by emily-bicks — COMMENTED** [review]
Looks good! my comments are mostly about how we want to organise the repo:

- Should the plotting scripts (i.e. histograms and sunbursts) go in analysis/validation?
- For the utils I feel like there are two separate categories in utils/validation.py and pipeline/make_histogram_levels.py. I think it may be more intutitive to break it into two scripts:
1. utils/frequency_distributions.py: this could be for all of the different frequency distributions that we're considering for the taxonomy (docs per tax group, entities per tax group, institutions per tax group, journals per tax group)
2. utils/validation/journal_validation.py: the more specific utils that are don't fall into the above category and are used for the journal validation
Just a suggestion!

**Comment by ampudia19** [comment]
In regards to the validation utils, I completely agree that it eventually needs splitting. In fact, most functions there correspond to the labelling approaches (and using it to label our taxonomy with entities or journals). I'll rename it to "labelling.py", and create other py files as per your request for subsequent work, if necessary.

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py (hunk 3 above)
Comment (ampudia19) [reply dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:22]: Probably a more generalised version, but I agree.

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py (hunk 4 above)
Comment (ampudia19) [reply dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:89]: Now that the pipelines are harmonised (changes in #9 to match the way you generated the dfs), yeah.

File: dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py (hunk 5 above)
Comment (ampudia19) [reply dap_aria_mapping/pipeline/taxonomy_validation/histograms/make_histogram_levels.py:142]: I already use the formatting tool (great one, btw!). Not sure what the returns from using the saving.save function are, over chart.save.

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py (hunk 6 above)
Comment (ampudia19) [reply dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: Yeah, and will create the same argument as in the histograms to allow a random number of taxonomies to be passed.

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py (hunk 7 above)
Comment (ampudia19) [reply dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: But this file simply creates labels (of either entity or journal types). It doesn't output the distributions.

File: dap_aria_mapping/utils/semantics.py (hunk 8 above)
Comment (ampudia19) [reply dap_aria_mapping/utils/semantics.py]: Sure! These are originally in the semantics utils in #9, hence me trying to keep them consistent across branches. But will just make sure I move them to entity_selection.

File: dap_aria_mapping/utils/validation.py (hunk 9 above)
Comment (ampudia19) [reply dap_aria_mapping/utils/validation.py]: Hopefully clearer now.

File: dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py (hunk 7 above)
Comment (emily-bicks) [reply dap_aria_mapping/pipeline/taxonomy_validation/make_labels.py]: Ah ok right - so by "label" you mean associate entities or journals with taxonomy groups? Maybe break it up into tag_groups_with_entities.py and tag_groups_with_journals.py ? Because I feel like down the road if we 'name' the taxonomy groups that will make the idea of 'labels' a bit confusing if they're different than the names.

**Comment by ampudia19** [comment]
@emily-bicks I pushed some additional code (relating to the histograms, mostly). I was thinking of directly merging to dev since you already reviewed its legacy code, but maybe you want to have a quick peak again. Otherwise I can merge :)

**Review by emily-bicks — APPROVED** [review]
